src/transformers/models/hiera/modeling_hiera.py

```diff
@@ -0,0 +1,1660 @@
+                dim_out=dim_out,
+                num_heads=config.num_heads[idx_stage],
+                drop_path=dpr[sum(config.depths[:idx_stage]) : sum(config.depths[: idx_stage + 1])],
+                query_stride=query_strides[sum(config.depths[:idx_stage]) : sum(config.depths[: idx_stage + 1])],
+                window_size=int(math.prod(config.masked_unit_size) * math.prod(config.query_stride) ** -idx_stage),
+                use_mask_unit_attn=config.masked_unit_attention[idx_stage],
+                stage_num=idx_stage,
+            )
+
+            embed_dim = dim_out
+            self.stages.append(stage)
+
+        # Setting reroll schedule
+        # The first stage has to reverse everything
+        # The next stage has to reverse all but the first unroll, etc.
+        stage_size = [i // s for i, s in zip(config.image_size, config.patch_stride)]
+        unroll_schedule = [config.query_stride] * len(config.depths[:-1])
+
+        self.schedule = {}
+        for idx_stage in range(len(config.depths)):
+            self.schedule[idx_stage] = unroll_schedule, stage_size
+            if idx_stage < config.num_query_pool:
+                stage_size = [i // s for i, s in zip(stage_size, config.query_stride)]
+                unroll_schedule = unroll_schedule[1:]
+
+        self.gradient_checkpointing = False
+
+    def reroll(
+        self, hidden_states: torch.Tensor, stage_idx: int, mask: Optional[torch.BoolTensor] = None
+    ) -> torch.Tensor:
+        """
+        Roll the given tensor back up to spatial order assuming it's from the given block.
+
+        If no mask is provided returns:
+            - [batch_size, height, width, hidden_size] for 2d
+            - [batch_size, frames, height, width, hidden_size] for 3d
+        If a mask is provided returns:
+            - [batch_size, num_mask_units, mask_unit_height, mask_unit_width, hidden_size] for 2d
+        """
+        schedule, size = self.schedule[stage_idx]
+        batch_size, seq_len, hidden_size = hidden_states.shape
+
+        num_dim = len(size)
+        mask_unit_shape = [1] * num_dim
+
+        for strides in schedule:
+            # Extract the current patch from seq_len
+            hidden_states = hidden_states.view(
+                batch_size, *strides, seq_len // math.prod(strides), *mask_unit_shape, hidden_size
+            )
+
+            # Move that patch into the current MU
+            # Example in 2d:
+            # Input: [batch_size, stride, stride, seq_len//(stride*stride), mask_unit_height, mask_unit_width, hidden_size]
+            # Output: [batch_size, seq_len//(stride*stride), stride, mask_unit_height, stride, mask_unit_width, hidden_size]
+            L = len(hidden_states.shape)
+            permute = (
+                [0, 1 + num_dim]
+                + sum(
+                    [list(p) for p in zip(range(1, 1 + num_dim), range(1 + num_dim + 1, L - 1))],
+                    [],
+                )
+                + [L - 1]
```

Comment:
This is a complex and non-obvious way to flatten a list. Let's use list -comprehensions which are more explicit and pythonic

```suggestion
                + [x for p for p in zip(range(1, 1 + num_dim), range(1 + num_dim + 1, L - 1)) for x in p],
                + [L - 1]
```

Reply:
We can remove this altogether actually if we plan to implement the video checkpoints as a separate implementation as this logic handles both image and video case